File: raspberry/SmartHome/app/src/main/java/smarthome/raspberry/server/WebServer.java
package smarthome.raspberry.server;

import android.support.annotation.NonNull;
import android.util.Log;

import java.io.IOException;


import fi.iki.elonen.NanoHTTPD;
public class WebServer extends NanoHTTPD implements StoppableServer {

    public static final String TAG = WebServer.class.getSimpleName();
    public static final int PORT = 8080;

    public WebServer() {
        super(PORT);
    }

    @Override
    public Response serve(IHTTPSession session) {
        try {
            return HandlerType.handle(session);
        } catch (Exception e) {
            Log.d(TAG, "can't serve: " + e);
            return HandlerType.errorHandle("No suitable method found");
        }
    }

    @Override
    public void startServer() {
        try {
            start();
        } catch (IOException e) {
            Log.e(TAG, "startServer: ", e);
        }
    }

    @Override
    public void stopServer() {
        stop();
    }
}

File: raspberry/SmartHome/app/src/main/java/smarthome/raspberry/server/WebServer.kt
package smarthome.raspberry.server

import android.util.Log

import java.io.IOException

import fi.iki.elonen.NanoHTTPD
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext

const val TAG = "WebServer"
const val PORT = 8080

class WebServer : NanoHTTPD(PORT), StoppableServer {
    override fun serve(session: NanoHTTPD.IHTTPSession): NanoHTTPD.Response {
        return try {
            val response = runBlocking { HandlerType.handle(session) }
            response
        } catch (e: Exception) {
            Log.d(TAG, "can't serve: $e")
            HandlerType.errorHandle("No suitable method found")
        }

    }

    override fun startServer() {
        try {
            start()
        } catch (e: IOException) {
            Log.e(TAG, "startServer: ", e)
        }

    }

    override fun stopServer() {
        stop()
    }
}
